packages/postcss/src/index.ts

```diff
@@ -21,47 +21,61 @@ export const loadConfig = () => interopDefault(customRequire('@pandacss/postcss'
 let stream: ReturnType<typeof setLogStream> | undefined
 
 const builder = new Builder()
+let builderGuard: Promise<void> | undefined
 
 export const pandacss: PluginCreator<PluginOptions> = (options = {}) => {
   const { configPath, cwd, logfile, allow } = options
 
   if (!stream && logfile) {
     stream = setLogStream({ cwd, logfile })
   }
+  const postcssProcess: TransformCallback = async function (root, result) {
+    const fileName = result.opts.from
 
-  return {
-    postcssPlugin: PLUGIN_NAME,
-    plugins: [
-      async function (root, result) {
-        const fileName = result.opts.from
+    const skip = shouldSkip(fileName, allow)
+    if (skip) return
 
-        const skip = shouldSkip(fileName, allow)
-        if (skip) return
+    await builder.setup({ configPath, cwd })
 
-        await builder.setup({ configPath, cwd })
+    // ignore non-panda css file
+    if (!builder.isValidRoot(root)) return
 
-        // ignore non-panda css file
-        if (!builder.isValidRoot(root)) return
+    await builder.emit()
 
-        await builder.emit()
+    builder.extract()
 
-        builder.extract()
+    builder.registerDependency((dep) => {
+      result.messages.push({
+        ...dep,
+        plugin: PLUGIN_NAME,
+        parent: result.opts.from,
+      })
+    })
 
-        builder.registerDependency((dep) => {
-          result.messages.push({
-            ...dep,
-            plugin: PLUGIN_NAME,
-            parent: result.opts.from,
-          })
-        })
+    builder.write(root)
 
-        builder.write(root)
+    root.walk((node) => {
+      if (!node.source) {
+        node.source = root.source
+      }
+    })
+  }
 
-        root.walk((node) => {
-          if (!node.source) {
-            node.source = root.source
-          }
+  return {
+    postcssPlugin: PLUGIN_NAME,
+    plugins: [
+      async function (...args) {
+        for (;;) {
+          if (!builderGuard) break
+          await builderGuard
+        }
+        const _builderGuard = postcssProcess(...args)
+        if (!_builderGuard) return
+
+        builderGuard = _builderGuard.finally(() => {
+          builderGuard = undefined
         })
+        return builderGuard
```

Comment:
what is happening here ? I'm not familiar with this `for (;;)` trick

also, would it be possible to add a test somewhere in there https://github.com/chakra-ui/panda/blob/main/packages/postcss/__tests__/postcss.test.ts

Reply:
Well, this is heavily based on javascript promises, async, and single threaded nature. Trying to explain:

- When postcss executes method first time, `builderGuard` is undefined, so `if (!builderGuard) break` will be executed, then processing starts `const _builderGuard = postcssProcess(...args)` (note, no await, so `_builderGuard` is pending promise). `if (!_builderGuard) return` is just to make typescrupt happy. We assign promise to `builderGuard` global variable and add `_builderGuard.finally` to resets this global when promise is settled. Then we return this promise to postcss, and it starts waiting this processing to be finished.
- If builder is still processing, and another call arrives - `builderGuard` will be pending promise, so `await builderGuard` is executed and it waits for processing to be finished. Multiple calls may wait for this promise simultaneously.
- If processing is finished, `builderGuard` will be settled and global variable will be `undefined`, and one of the `await builderGuard` will continue. Because `builderGuard` is now undefined (in `finally`) cycle will break and another processing will start. All other simultaneous tasks will still wait, because of this `for (;;)` loop.

I added simple test, and it works, but it requires exposing builder instance. `export const builder`

Reply:
got it, so basically you're waiting from a global promise

could this work ? it seems a bit simpler
```ts
 async function (...args) {
      if (builderGuard) await builderGuard.catch(() => {});
      builderGuard = postcssProcess(...args);
      return builderGuard;
    },
```

Reply:
This variant will fix concurrent shared context initialization bug.

I think it will work correctly if there is only one concurrent invocation waiting. I.e. one waiting and one processing. If there are two waiting (in my project there were 3 concurrent invocations) - they will wait for first task to complete, and then they both will be executed concurrently.

It can work assuming there will be no other concurrent invocation bugs in builder, but to be on a safe side I'd rewrite it to something like:

```javascript
      function (...args) {
        builderGuard = Promise.resolve(builderGuard)
          .catch(() => {})
          .then(() => postcssProcess(...args))
        return builderGuard
      },
```

There are no await between reading and setting `builderGuard` so it will always point to last executed process.

Reply:
PR updated